### PR TITLE
[RFC] Standardize CustomCallOp to extend backend_config to take a DictionaryAttr

### PR DESCRIPTION
**Pull request by abhigunj**
This RFC proposes standardizing CustomCallOp to extend backend_config to take a DictionaryAttr.

Please review and provide you feedback.

### CODE REVIEW

**Review by mlevesquedion — APPROVED** [review]

File: rfcs/20240312-standardize-customcallop.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+Several features have been added to MHLO in the past year, which frameworks want
+to leverage and members of the community have made requests for them as well.
+This includes: MHLO
+[`custom_call`](https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/IR/hlo_ops.td#L2483)
+`backend_config` to take a `DictionaryAttr`.
+
+StableHLO `custom_call` op does not support this feature. There are several
+occurrences of users working around this gap in `custom_call` today, examples -
+Jax uses [unregistered attributes](https://github.com/google/jax/blob/1ed27ecebb92e916b45601e3a107971170a4592b/jaxlib/hlo_helpers.py#L191)
```
Comment (ghpvnist) [inline rfcs/20240312-standardize-customcallop.md:18]: Nit: JAX

**Review by ghpvnist — APPROVED** [review]
LGTM with minor comments

File: docs/spec.md [hunk 2]
```diff
@@ -2437,8 +2437,9 @@ the XLA compiler. In the future, we are planning to unify this metadata
 %results = "stablehlo.custom_call"(%input0) {
   call_target_name = "foo",
   has_side_effect = false,
-  backend_config = "bar",
-  api_version = 1 : i32,
+  backend_config = {bar = 42 : i32},
```
[the remark this answers is not in the document]
Comment (mlevesquedion) [reply docs/spec.md:2440]: I'm pretty sure this is correct? MLIR doesn't insert spaces between the brackets and the content:

```
$ stablehlo-opt <<<"func.func @foo() {
  stablehlo.composite \"foo.op\" {
    composite_attributes = { bar = 42 : i32 },
    decomposition = @foo
  } : () -> ()
  func.return
}"""
module {
  func.func @foo() {
    stablehlo.composite "foo.op"  {composite_attributes = {bar = 42 : i32}, decomposition = @foo} : () -> ()
    return
  }
}
```
(this uses composite but the idea is the same)

[the remark this answers is not in the document]
Comment (ghpvnist) [reply docs/spec.md:2440]: Hm you're right, looks like our spec needs some updates. I'll create a separate PR to fix them, you can ignore this comment!

**Review by sdasgup3 — APPROVED** [review]

**Review by GleasonK — CHANGES_REQUESTED** [review]
Adding comment to block until we have community alignment / a chance to socialize this further internally+externally

File: rfcs/20240312-standardize-customcallop.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+## Motivation
+
+Several features have been added to MHLO in the past year, which frameworks want
+to leverage and members of the community have made requests for them as well.
+This includes: MHLO
+[`custom_call`](https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/IR/hlo_ops.td#L2483)
+`backend_config` to take a `DictionaryAttr`.
+
+StableHLO `custom_call` op does not support this feature. There are several
```
Comment (cantonios) [inline rfcs/20240312-standardize-customcallop.md:16]: nit: "The current" StableHLO `custom_call`

File: rfcs/20240312-standardize-customcallop.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+occurrences of users working around this gap in `custom_call` today, examples -
+JAX uses [unregistered attributes](https://github.com/google/jax/blob/1ed27ecebb92e916b45601e3a107971170a4592b/jaxlib/hlo_helpers.py#L191)
+to hold the `DictionaryAttr` or serializing the dictionary as a string
+to pass around for StableHLO `custom_call` op.
```
Comment (cantonios) [inline rfcs/20240312-standardize-customcallop.md:20]: nit: "the" StableHLO

File: rfcs/20240312-standardize-customcallop.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+We propose standardizing StableHLO `custom_call` op to extend `backend_config`
```
Comment (cantonios) [inline rfcs/20240312-standardize-customcallop.md:22]: nit: "the" StableHLO `custom_call` op

File: rfcs/20240312-standardize-customcallop.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+to take a `DictionaryAttr`.  So they can be used safely by the community without
```
Comment (cantonios) [inline rfcs/20240312-standardize-customcallop.md:23]: nit: to take a `DictionaryAttr` so they can be used ...

File: rfcs/20240312-standardize-customcallop.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+workarounds. It will help to unify metadata under single `DictionaryAttr` which
```
Comment (cantonios) [inline rfcs/20240312-standardize-customcallop.md:24]: nit: It -> This

Comment (cantonios) [inline rfcs/20240312-standardize-customcallop.md:24]: nit: under "a" single `DictionaryAttr`

**Review by GleasonK — APPROVED** [review]

**Pull request merged**